src/xenia/kernel/xam/apps/xmp_app.cc

```diff
@@ -30,7 +47,240 @@ XmpApp::XmpApp(KernelState* kernel_state)
+      xe::vfs::FileAccess::kGenericRead, false, true, &vfs_ctx.file,
+      &file_action);
+  if (XFAILED(status)) {
+    XELOGE("Opening {} failed with status {:X}", filename, status);
+    return false;
+  }
+
+  AVIOContext* avio_ctx = avio_alloc_context(buffer, buffer_size, 0, &vfs_ctx,
+                                             xenia_vfs_read, nullptr, nullptr);
+
+  AVFormatContext* formatContext = avformat_alloc_context();
+  formatContext->pb = avio_ctx;
+  int ret;
+  if ((ret = avformat_open_input(&formatContext, nullptr, nullptr, nullptr)) !=
+      0) {
+    XELOGE("ffmpeg: Could not open WMA file: {:x}", ret);
+    av_freep(&avio_ctx->buffer);
+    avio_context_free(&avio_ctx);
+    return false;
+  }
+
+  if (avformat_find_stream_info(formatContext, nullptr) < 0) {
+    XELOGE("ffmpeg: Could not find stream info");
+    avformat_close_input(&formatContext);
+    av_freep(&avio_ctx->buffer);
+    avio_context_free(&avio_ctx);
+    return false;
+  }
+
+  AVCodec* codec = nullptr;
+  int streamIndex =
+      av_find_best_stream(formatContext, AVMEDIA_TYPE_AUDIO, -1, -1, &codec, 0);
+  if (streamIndex < 0) {
+    XELOGE("ffmpeg: Could not find audio stream");
+    avformat_close_input(&formatContext);
+    av_freep(&avio_ctx->buffer);
+    avio_context_free(&avio_ctx);
+    return false;
+  }
+
+  AVStream* audioStream = formatContext->streams[streamIndex];
+  AVCodecContext* codecContext = avcodec_alloc_context3(codec);
+  avcodec_parameters_to_context(codecContext, audioStream->codecpar);
+
+  if (audioStream->codecpar->format != AV_SAMPLE_FMT_FLTP &&
+      audioStream->codecpar->format != AV_SAMPLE_FMT_FLT) {
+    XELOGE("Audio stream has unexpected sample format {:d}",
+           audioStream->codecpar->format);
+    avcodec_free_context(&codecContext);
+    avformat_close_input(&formatContext);
+    av_freep(&avio_ctx->buffer);
+    avio_context_free(&avio_ctx);
+    return false;
+  }
+
+  if (avcodec_open2(codecContext, codec, nullptr) < 0) {
+    XELOGE("ffmpeg: Could not open codec");
+    avcodec_free_context(&codecContext);
+    avformat_close_input(&formatContext);
+    av_freep(&avio_ctx->buffer);
+    avio_context_free(&avio_ctx);
+    return false;
+  }
+
+  auto driverReady = xe::threading::Semaphore::Create(64, 64);
+  {
+    std::unique_lock<std::mutex> guard(driver_mutex_);
+    driver_ = kernel_state_->emulator()->audio_system()->CreateDriver(
+        driverReady.get(), codecContext->sample_rate, codecContext->channels,
+        false);
+    if (!driver_->Initialize()) {
+      XELOGE("Driver initialization failed!");
+      driver_->Shutdown();
+      driver_ = nullptr;
+      avcodec_free_context(&codecContext);
+      avformat_close_input(&formatContext);
+      av_freep(&avio_ctx->buffer);
+      avio_context_free(&avio_ctx);
+      return false;
+    }
+  }
+  if (volume_ == 0.0f) {
+    // Some games set volume to 0 on startup and then never call SetVolume again...
+    volume_ = 0.7f;
```

Comment:
This should be 1.0f (100%) instead. That's the default value on console and xenia. Also, it would be nice to know why it happens in the first place and fix it properly. This feels more like a patch for the actual issue/behavior.

Reply:
I disagree that it should be 100%. That's a comparatively loud value and it pretty much drowns out most other sounds of a game. Your point that this is pretty much a bandaid is valid, I'd also like to know why the game does this and won't attempt to re-set the volume at a later point.

Reply:
So, you are saying that having a randomly chosen value is more "accurate"? Also, you initialize the volume variable at 1.0f but decided to set a random value when this issue happens. It's not consistent... not only in this implementation but with the audio volume in xenia, which also initializes at 1.0f. And if the music is too loud, you can always adjust it in the game settings. Or make it a configurable variable and put it in the config. At least, users can adjust it whatever they feel like it.

Reply:
I made it configurable now. The constructor code wasn't written by me - I've adjusted it to be consistent. I do believe that the "randomly chosen value" reflects real world behavior more accurately.

Unfortunately the ingame volume settings have no effect in PGR3 either; the game is not making SetVolume calls to apply the changes you make in the settings menu. Perhaps an explanation would be that it's trying to tell the system about the audio settings through some other non-XMP mechanism? I don't know what that would be, though.

Reply:
Oh well, I guess is good enough for now. Maybe Gliniak can find a solution for both issues in the future. Can we make the values in the config to be from 0 to 10 instead? It's more user friendly. Thanks.

Reply:
Would you be fine with 0-100? Seems more natural because it's a percentage.

Reply:
> Would you be fine with 0-100? Seems more natural because it's a percentage.

You would need to divide it by 100 so it stays between 0.0-1.0 (at least with xaudio): https://learn.microsoft.com/en-us/windows/win32/xaudio2/how-to--change-voice-volume